### DISCUSSION TITLE
Webpack `cache.buildDependencies` dependencies resolver works incorectly.

### DISCUSSION THREAD

**Discussion by zdm** [webpack/webpack#17176] [Q&A]
# Bug report

**What is the current behavior?**

**If the current behavior is a bug, please provide the steps to reproduce.**

My `webpack congig` depends on `package-a`.

`package-a` is depends on `package-b`;

For development purposes packages `package-a` and `package-b` symlinked to the `node_modules`.

I set `NODE_PRESERVE_SYMLINKS=1`, so nodejs will not resolves symlinks.

I added my `webpack congig` location to the `cache.buildDependencies`.

Whe I am running build process I got error:

`package-b` can't be resolved from the `package-a`.
And in error message it shows original `package-a` location (not symlink location).

So, seems, that when webpack resolves `cache.buildDependencies` dependencies it does not respect `NODE_PRESERVE_SYMLINKS=1` and follwos symbolic links.

This disables filesysten cache.

**What is the expected behavior?**

I am expecting, that when `NODE_PRESERVE_SYMLINKS=1` is set - modules resilver will not follw symbolic links. As it implemented in the nodejs.

**Other relevant information:**
webpack version:  5.82.0
Node.js version: 20.0.1
Operating System: win / linux
Additional tools: none

**Comment by alexander-akait** [comment]
Sorry, we don't support `NODE_PRESERVE_SYMLINKS` and I think we will to support this, please avoid using this, it is for Node.js CLI, we have https://webpack.js.org/configuration/resolve/#resolvesymlinks, please use this

**Comment by zdm** [comment]
Thank you for reply.
I added `resolve.symlinks = false` to my config.
But nothing was changed, same error occures, cache not created.
Does this option works?

**Comment by zdm** [comment]
Semms, that this option is for stage, when webpack build sources, but not for `cache.buildDependencies`.
I am right?

**Comment by zdm** [comment]
Remember, that my case is for webpack config dependencies.

No options to fix this?

**Comment by zdm** [comment]
@alexander-akait 
Sorry for disturbing you, I know that you are very busy, but vould you please help me to understand why this option is not work in my case?

**Comment by zdm** [comment]
@alexander-akait 
Nothing works for me, I see this error and unable to use cache.
I there is a way to debug `enhanced-resolver`?
I want to solve this problem.

```
<w> [webpack.cache.PackFileCacheStrategy] Caching failed for pack: Error: Can't resolve '@softvisio/vue/webpack-components/main' in 'd:\projects\softvisio-node\vue-ext\resources\webpack-components\main'
<w> while resolving '@softvisio/vue/webpack-components/main' in d:\projects\softvisio-node\vue-ext\resources\webpack-components\main as file
<w>  at resolve esm file @softvisio/vue/webpack-components/main
<w>  at file dependencies d:\projects\softvisio-node\vue-ext\resources\webpack-components\main\component.js
<w>  at file d:\projects\softvisio-node\vue-ext\resources\webpack-components\main\component.js
<w>  at resolve esm file @softvisio/vue-ext/webpack-components/main
<w>  at file dependencies d:\projects\zdm\playground\frontend\resources\webpack-components\main\component.js
<w>  at file d:\projects\zdm\playground\frontend\resources\webpack-components\main\component.js
<w>  at resolve commonjs d:\projects\zdm\playground\frontend\resources\webpack-components\main\component.js
<w> [webpack.cache.PackFileCacheStrategy] Caching failed for pack: Error: Can't resolve '@softvisio/vue/webpack-components/main' in 'd:\projects\softvisio-node\vue-ext\resources\webpack-components\main'
<w> while resolving '@softvisio/vue/webpack-components/main' in d:\projects\softvisio-node\vue-ext\resources\webpack-components\main as file
<w>  at resolve esm file @softvisio/vue/webpack-components/main
<w>  at file dependencies d:\projects\softvisio-node\vue-ext\resources\webpack-components\main\component.js
<w>  at file d:\projects\softvisio-node\vue-ext\resources\webpack-components\main\component.js
<w>  at resolve esm file @softvisio/vue-ext/webpack-components/main
<w>  at file dependencies d:\projects\zdm\playground\frontend\resources\webpack-components\main\component.js
<w>  at file d:\projects\zdm\playground\frontend\resources\webpack-components\main\component.js
<w>  at resolve commonjs d:\projects\zdm\playground\frontend\resources\webpack-components\main\component.js
```

**Reply by alexander-akait** [reply]
Looks like the problem in `@softvisio/vue/webpack-components/main` resolving, if you disable cache all works fine?

**Reply by alexander-akait** [reply]
You can create reproducible test repo I will help

**Comment by zdm** [comment]
Hi.

Reproduction is ready.
I think the problem is that `extended-resolver` does not take in account `node --preserve-symlinks` option and unable to resolve dependencies.
Please, help me to solve this.

### How to reproduce

```shell
# clone repo
git clone https://github.com/zdm/webpack-build-dependencies-issue-17169

cd webpack-build-dependencies-issue-17169
cd main

# this command:
# - installs dependencies
# - links "package-a" and "package-b" to the "main/node_modules"
npm i

# this command runs:
# node --preserve-symlinks node_modules/webpack/bin/webpack.js
npm run build
```

Build is OK, but cache is not used due to the following errors:

```txt
<w> [webpack.cache.PackFileCacheStrategy] Caching failed for pack: Error: Can't resolve 'package-b/webpack.config.js' in 'd:\projects\devel\issues\webpack-build-dependencies-issue-17169\package-a'
<w> while resolving 'package-b/webpack.config.js' in d:\projects\devel\issues\webpack-build-dependencies-issue-17169\package-a as file
<w>  at resolve esm file package-b/webpack.config.js
<w>  at file dependencies d:\projects\devel\issues\webpack-build-dependencies-issue-17169\package-a\webpack.config.js
<w>  at file d:\projects\devel\issues\webpack-build-dependencies-issue-17169\package-a\webpack.config.js
<w>  at resolve esm file package-a/webpack.config.js
<w>  at file dependencies d:\projects\devel\issues\webpack-build-dependencies-issue-17169\main\webpack.config.js
<w>  at file d:\projects\devel\issues\webpack-build-dependencies-issue-17169\main\webpack.config.js
<w>  at resolve commonjs d:\projects\devel\issues\webpack-build-dependencies-issue-17169\main\webpack.config.js
..
```

**Comment by zdm** [comment]
Seems, that `extended-resolver` is configured to resolve dependencies for sub-packages starting from the sub-package root dir, but the correct way is to resolve them starting from the main module root dir.

**Reply by alexander-akait** [reply]
I want to say it is a limitation, we never support `--preserve-symlinks`, also you packages use each other but do not declare it in their dev dependencies, technically it is possible to implement the possibility of configuring internal revolvers, but I'm not sure it will be worth the effort (very exotic case). I would recommend using - https://docs.npmjs.com/cli/v8/using-npm/workspaces, this is a more supported and official way to work with multiple packages in one repository.

**Comment by zdm** [comment]
This is not so rare case, this is very helpful when you for example need to work both on the end-user application and the shared web components libraries.

Interesting, that deps resolver failed only for build dependencies, but works for the same dependencies in the main build process.
Maybe fix can be simple?

**Reply by alexander-akait** [reply]
> This is not so rare case, this is very helpful when you for example need to work both on the end-user application and the shared web components libraries.

Can you show me other projects? I still recommend to use https://docs.npmjs.com/cli/v8/using-npm/workspaces

Your dependency tree is not quite correctly built, which can lead to extremely difficult and complex bugs.

https://nodejs.org/api/cli.html#--preserve-symlinks

From docs:

> Note, however, that using --preserve-symlinks can have other side effects. Specifically, symbolically linked native modules can fail to load if those are linked from more than one location in the dependency tree (Node.js would see those as two separate modules and would attempt to load the module multiple times, causing an exception to be thrown).

> Interesting, that deps resolver failed only for build dependencies, but works for the same dependencies in the main build process.
Maybe fix can be simple?

There is no concept of fix here, we just follow default resolution node.js logic

 > technically it is possible to implement the possibility of configuring internal revolvers, but I'm not sure it will be worth the effort (very exotic case)
 
 Here is place https://github.com/webpack/webpack/blob/main/lib/FileSystemInfo.js#L1513

**Comment by zdm** [comment]
1. I am unable to show you other projects, but any company or freelancer, who, for example, develop web aps, has own web components libraries or other packages with the shared code and this is very useful to have possibility to link this packages during development, this will simplify development and debugging. This is not a my personal invention.
2. workspaces are for multipackage repos, and will not help is this case.
3. the problem exactly in how you resolves links, you are not completely follow the specification, because node itself resolves them without any problems.
4. i will try to debug and will tell you.

**Comment by zdm** [comment]
`ebhanced-resolver` is too complex.

Did you considered to use built-in node resolcer.

```
import module from "node:module";

module.createRequire( from ).resolve( path );
```

It will work 100% correctly.

Only need to add support for custom aliases and some other features, which are used in webpack.

I can create the code.

**Reply by alexander-akait** [reply]
> the problem exactly in how you resolves links, you are not completely follow the specification, because node itself resolves them without any problems.

You can send a pr to `enhanced-resolve` 

> Did you considered to use built-in node resolcer.

No, because it does not fit our cache logic

**Reply by alexander-akait** [reply]
> workspaces are for multipackage repos, and will not help is this case

Just set valid `devDeps` and `deps` and use them as separate packages

**Comment by zdm** [comment]
Is it possible to access resolver, which is used to resolve build deps? I want to try to add plugin.

**Comment by zdm** [comment]
Easiest way is to ask Trump to make webpack great again.

**Comment by alexander-akait** [comment]
We can do it, I doubt it only because all the time using webpack you are the first and only one who has encountered this...

>  technically it is possible to implement the possibility of configuring internal revolvers, but I'm not sure it will be worth the effort (very exotic case)

I honestly can't understand what the problem is with using dependencies using `package.json` (`dependencies` and `devDependencies`), you don't specify dependencies directly, so they don't appear in `node_modules` and as i said earlier such behavior can lead to unpredictable results, for example duplicated packets can be considered as completely different, which will make your `bundle` bigger. 

I want to help you, I want to say that's such behavior is extremely undesirable and you could potentially get into serious trouble - your dependency tree is broken. The problem will be especially evident with transitive dependencies. And I don’t understand why you ignore the direct words from the author of the project that this approach is dangerous. 

As well as direct words from the developer of the node:

> Note, however, that using --preserve-symlinks can have other side effects. Specifically, symbolically linked native modules can fail to load if those are linked from more than one location in the dependency tree (Node.js would see those as two separate modules and would attempt to load the module multiple times, causing an exception to be thrown).

**Reply by alexander-akait** [reply]
We can implement support `NODE_PRESERVE_SYMLINKS` in `enhanced-resolve`, that's why I answered you earlier that

> You can send a pr to enhanced-resolve

Because there are no problems at all on the webpack side

**Comment by zdm** [comment]
In my reproduction I didn't specified dependencies, because package-a and package-b are not published.
In my other projects all deps are specified and results are the same.

Generally - it works, project compiled, but without cache. I am unable to be happy, when something not works.
Ok, I will debug it when will find a time, currently I am overloaded with the ongoing work.
Thank you for your help.

**Reply by alexander-akait** [reply]
> In my reproduction I didn't specified dependencies, because package-a and package-b are not published.

you can use them without publication - https://docs.npmjs.com/cli/v10/configuring-npm/package-json#local-paths

**Comment by zdm** [comment]
@alexander-akait 

Is it possible to use hooks for resolvers that are used to resolve build dependencies?
This problem can be fixed if we add something like `SymlinksPlugin` for this resolver.

**Reply by alexander-akait** [reply]
I don't think, but I am fine to add option(s) for this

**Comment by zdm** [comment]
Please add, I you can.
But at first, could you please point me to the place in the code, where build dependencies resolvers created?
I will test if symlink resolver will help.

**Reply by alexander-akait** [reply]
Feel free to send a PR, it is exotic case, so it doesn't have high priority 

Here is the place where they are creating - https://github.com/webpack/webpack/blob/main/lib/FileSystemInfo.js#L1541
